src/ray/core_worker/core_worker.h

```diff
@@ -59,6 +59,99 @@ namespace core {
 
 JobID GetProcessJobID(const CoreWorkerOptions &options);
 
+/// Simple container for per function task counters. The counters will be
+/// keyed by the function name in task spec.
+struct TaskCounter {
+  /// A task can only be one of the following state. Received state in particular
+  /// covers from the point of RPC call to beginning execution.
+  enum TaskStatusType { kPending, kRunning, kFinished };
+
+  /// This mutex should be used by caller to ensure consistency when transitioning
+  /// a task's state.
+  mutable absl::Mutex tasks_counter_mutex_;
+  absl::flat_hash_map<std::string, int> pending_tasks_counter_map_
+      GUARDED_BY(tasks_counter_mutex_);
+  absl::flat_hash_map<std::string, int> running_tasks_counter_map_
+      GUARDED_BY(tasks_counter_mutex_);
+  absl::flat_hash_map<std::string, int> finished_tasks_counter_map_
+      GUARDED_BY(tasks_counter_mutex_);
+
+  /// Tracks the total number of running tasks on this process.
+  int64_t running_total_ GUARDED_BY(tasks_counter_mutex_) = 0;
```

Comment:
This is getting a bit complex; but in a follow-up PR, I have a cleanup that refactors out a common Counter to simplify across the different locations we're counting tasks by key.